Fix <C-S> during insert mode on mswin.vim

abangfarhan opened the pull request:
Originally, the mapping for `<C-S>` on mswin.vim during insert mode is:

```
inoremap <C-S> <C-O>:update<CR>
```

This mapping will cause problem during omnicompletion. Omnicompletion is triggered by `<C-X><C-O>`, and so if you type `<C-O>` during omnicompletion it will not change to normal mode, but cycle through the suggestions.

abangfarhan committed:
Fix <C-S> during insert mode on mswin.vim

Originally, the mapping for <C-S> on mswin.vim during insert mode is:

inoremap <C-S> <C-O>:update<CR>

This mapping will cause problem during omnicompletion. Omnicompletion is triggered by <C-X><C-O>, and so if you type <C-O> during omnicompletion it will not change to normal mode, but cycle through the suggestions.

brammool commented:
I don't understand the problem you are fixing.  CTRL-S is changed, but you mention a problem with CTRL-O

abangfarhan commented:
I will give an example. Suppose you want to edit an HTML file with the following minimal vimrc:

```
source $VIMRUNTIME/mswin.vim
set nocompatible
set omnifunc=htmlcomplete#CompleteTags
```

Then you type the following:

```
<head></
```

Then, while still in insert mode, type in `<C-X><C-O>` to trigger omnicompletion. The tag will be completed. Then type `<C-S>`. The text will change to this:

```
<head></:update

```

Because typing `<C-S>` is mapped to `<C-O>:update<CR>`, when you use `<C-S>` during omnicompletion what happens is that the suggestions will be cycled once, insert the word `:update`, then insert a new line (instead of exit insert mode, save the file, then go back to insert mode).

Note that if this fix could not be implemented for whatever reason, then anyone who get annoyed by this problem can simply remap the `<C-S>` in their own vimrc:

```
inoremap <C-S> <Esc>:update<CR>a
```

tonymec commented:
My opinion about this kind of problem has always been: keep away from mswin.vim, even avoid it like the plague, its mappings hide a lot of useful Vim key bindings which would be better left at their original meaning. You just found one more example of that.

Or else (but that's not my way of thinking) avoid the vanilla-Vim key bindings which mswin.vim has hidden, or create `:noremap` mappings with them as `{rhs}` to save them from oblivion.

IMHO the moral of the story is: You can't have your cake and eat it too.

Best regards,
Tony.

chrisbra commented:
I think instead of using `<C-O>` to switch to normal mode for a simple command, we might rather use `<C-\><C-N>` which should work in any mode. However we need to re-enter insert mode afterwards.

brammool commented:
> I will give an example. Suppose you want to edit an HTML file with the following minimal vimrc:
> 
> ```
> source $VIMRUNTIME/mswin.vim
> set nocompatible
> set omnifunc=htmlcomplete#CompleteTags
> ```
> 
> Then you type the following:
> 
> ```
> <head></
> ```
> 
> Then, while still in insert mode, type in `<C-X><C-O>` to trigger omnicompletion. The tag will be completed. Then type `<C-S>`. The text will change to this:
> 
> ```
> <head></:update
> 
> ```
> 
> Because typing `<C-S>` is mapped to `<C-O>:update<CR>`, when you use `<C-S>` during omnicompletion what happens is that the suggestions will be cycled once, insert the word `:update`, then insert a new line (instead of exit insert mode, save the file, then go back to insert mode).
> 
> Note that if this fix could not be implemented for whatever reason, then anyone who get annoyed by this problem can simply remap the `<C-S>` in their own vimrc:
> 
> ```
> inoremap <C-S> <Esc>:update<CR>a
> ```

I think you need to use "gi" instead of "a".  Try it out at the start
and end of a line.

-- 
Facepalm statement #2: "If there is a country without immigrants I'm going to
move there"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

abangfarhan committed:
Changed `a` to `gi` in `inoremap <C-S>`

With `a`, using `<C-S>` on insert mode will move the cursor one character to theright when you are saving at the beginning of a nonempty line. Using `gi` will avoid this problem.

brammool commented:
OK, I'll include that.
